corehq/apps/enterprise/api/resources.py

```diff
@@ -0,0 +1,315 @@
+
+from django.http import HttpResponse, HttpResponseForbidden
+from django.template.loader import render_to_string
+from django.utils.translation import gettext as _
+
+from dateutil import tz
+from tastypie import fields
+from tastypie.exceptions import ImmediateHttpResponse
+
+from corehq.apps.accounting.models import BillingAccount
+from corehq.apps.accounting.utils.account import (
+    get_account_or_404,
+    request_has_permissions_for_enterprise_admin,
+)
+from corehq.apps.api.odata.utils import FieldMetadata
+from corehq.apps.api.odata.views import add_odata_headers
+from corehq.apps.api.resources import HqBaseResource
+from corehq.apps.api.resources.auth import ODataAuthentication
+from corehq.apps.enterprise.enterprise import (
+    EnterpriseDomainReport,
+    EnterpriseFormReport,
+    EnterpriseMobileWorkerReport,
+    EnterpriseODataReport,
+    EnterpriseWebUserReport,
+)
+
+
+class EnterpriseODataAuthentication(ODataAuthentication):
+    def is_authenticated(self, request, **kwargs):
+        authenticated = super().is_authenticated(request, **kwargs)
+        if authenticated is not True:
+            return authenticated
+
+        domain = kwargs['domain'] if 'domain' in kwargs else request.domain
+        account = get_account_or_404(domain)
+        if not request_has_permissions_for_enterprise_admin(request, account):
+            raise ImmediateHttpResponse(
+                HttpResponseForbidden(_(
+                    "You do not have permission to view this feed."
+                ))
+            )
+
+        return True
+
+
+class ODataResource(HqBaseResource):
+    class Meta:
+        include_resource_uri = False
+        collection_name = 'value'
+        authentication = ODataAuthentication()
+        limit = 2000
+        max_limit = 10000
+
+    def alter_list_data_to_serialize(self, request, data):
+        result = super().alter_list_data_to_serialize(request, data)
+
+        path = urljoin(request.get_full_path(), 'schema/#feed')
+        result['@odata.context'] = request.build_absolute_uri(path)
+
+        meta = result['meta']
+        result['@odata.count'] = meta['total_count']
+        if 'next' in meta and meta['next']:
+            result['@odata.nextLink'] = request.build_absolute_uri(meta['next'])
+
+        del result['meta']
+        return result
+
+    def determine_format(self, request):
+        # Currently a hack to force JSON. XML is supported by OData, but "Control Information" fields
+        # (https://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#_Toc38457735)
+        # are not valid attribute names in XML -- instead, they need to be implemented on the object's element.
+        # https://docs.oasis-open.org/odata/odata-atom-format/v4.0/cs02/odata-atom-format-v4.0-cs02.html#_Entity
+        # provides an example.
+        # In the interrim, it seems forcing JSON to avoid XML is the best option
+        return 'application/json'
```

Comment:
Yes -- the default format is actually specified to be 'application/json' from TastyPie's `ResourceOptions`.  However, the built-in `determine_format` only falls back on that default if no better format is found. First, it looks for an explicitly requested format via the `format` request parameter. Otherwise, it looks at the `HTTP_ACCEPT` header and tries to get a best-fit for the acceptable formats. Only if it can't find any matches that resemble the accepted options does it fall back to the default of `application/json`.
Early in my testing, PowerBI seemed to be rapidly requesting the same resource multiple times, with different accepted formats. Eventually, it would request something that was closest fit to XML, and then an exception would occur which caused the PowerBI request to fail.

From my recent testing, there are no longer extraneous requests, so we might be able to get away with dropping this approach. That said, I could see the argument to just coerce 'xml' to 'json' if XML is requested, rather than throwing the exception. It's hacky, but I'm still not sure when PowerBI decides to sent XML accept headers, and the API currently doesn't support XML at all. Rather than hoping PowerBI will move on and attempt a different format, throwing an Exception will instead stop the process entirely.

Reply:
Not sure I follow entirely, but are you saying that PowerBI will request XML, but still accept JSON?  And saying "sorry, we don't do XML" won't cut it?  As in, PowerBI has weird/bad behavior and we're hacking around it here?

Reply:
Something like that. If doesn't explicitly request XML. At some point, it was providing a list of 'HTTP_ACCEPT' formats that wasn't exactly matching anything, but was most closely matching (by the tastypie algorithm) to XML. If you returned JSON from that call, PowerBI would be fine with it, but XML would cause an error on our side